# 딥러닝 기초 수학 실습

## 1. 선형회귀 : 평균 제곱근 오차(RMSE; Root Mean Square Error)
### 1) 선형 회귀란? 
- 1) 임의의 직선(y = ax + b)을 그어 
- 2) 이에 대한 평균 제곱근 오차를 구하고 
- 3) 이 값을 가장 작게 만들어 주는 a와 b 값을 찾아가는 작업

### 2) 평균제곱근 오차(RMSE)를 이용한 선형회귀 수정 방법
#### 여러 개의 입력 값을 계산하는 방법
- 1) 임의의 선을 그리고 난 후 
- 2) 이 선이 얼마나 잘 그려졌는지를 평가하여 
- 3) 조금씩 수정해 가는 방법을 사용

#### 그려진 선의 오차를 평가하는 오차 평가 알고리즘이 필요
- 평균제곱근 오차(rmse; root mean square error)

#### 일단 그리고 조금씩 수정해 나가기 방식 
- 가설을 하나 세운 뒤 이 값이 주어진 요건을 충족하는지를 판단하여 조금씩 변화를 주고, 
- 이 변화가 긍정적이면 오차가 최소가 될 때까지 이 과정을 계속 반복하는 방법

#### 나중에 그린 선이 먼저 그린 선보다 더 좋은 지 나쁜지를 판단하기 위해 필요한 것은?
- 각 선의 오차를 계산 --> 오차가 작은 쪽으로 바꾸는 알고리즘이 필요

In [1]:
# 경고 무시
from warnings import filterwarnings
filterwarnings("ignore")

## 2. 평균제곱근(RMSE) 오차 실습

In [2]:
import numpy as np

# 기울기 a와 y 절편 b 정하기
ab = [3, 76]  # a = 3, b = 76

# x, y의 데이타 값 [공부시간, 점수]
data = [[2, 81], [4, 93], [6, 91], [8, 97]]
x = [i[0] for i in data] # x - 공부시간
y = [i[1] for i in data] # y - 점수

# y = ax+b에 a와 b 값을 대입하여 결과를 출력하는 함수
def predict(x):
    return ab[0]*x + ab[1]

# 평균제곱근(RMSE) 오차 함수
def rmse(p, a):
    return np.sqrt(((p - a)**2).mean())

# 평균제곱근(RMSE) 함수를 각 y값에 대입하여 최종값을 구하는 함수
def rmse_val(priedict_result, y):
    return rmse(np.array(predict_result), np.array(y))

# 예측값이 들어갈 빈 리스트
predict_result = []

# 모든 x값을 한번씩 대입하여 predict_result 리스트 완성
for i in range(len(x)):
    predict_result.append(predict(x[i]))
    print("공부한 시간 : %.f, 실제 점수 : %.f, 예측 점수 : %4.1f" % (x[i], y[i], predict(x[i])))
    
# 최종 RMSE 출력
print("RMSE 최종값 : " + str(rmse_val(predict_result, y)))


공부한 시간 : 2, 실제 점수 : 81, 예측 점수 : 82.0
공부한 시간 : 4, 실제 점수 : 93, 예측 점수 : 88.0
공부한 시간 : 6, 실제 점수 : 91, 예측 점수 : 94.0
공부한 시간 : 8, 실제 점수 : 97, 예측 점수 : 100.0
RMSE 최종값 : 3.3166247903554


### 중간 결론
- 기울기 a = 3, 절편 b = 76 --> 오차가 약 3.3166이라는 것을 알게 됨
- 이제 남은 것은 이 오차를 줄이면서 새로운 선을 긋는 것
- 이를 위해서는 a와 b의 값을 적절히 조절하면서 오차의 변화를 살펴보고, 
- 그 오차가 최소화되는 a와 b의 값을 구해야 함(미분 필요)
- 오차 수정하기 --> "경사하강법"